### DISCUSSION TITLE
Running zjs-ui in docker on pimox LXC: no serial device passthrough

### DISCUSSION THREAD

**Discussion by Wri84** [zwave-js/zwave-js-ui#2738] [Q&A] [answered]
Hi,

I'm trying to run zwave-js-ui inside an debian LXC container with docker on pimox on an raspberry pi. I know, complicated setup...

When running zjs-ui in docker baremetal, I'm able to passthrough my Z-Pi7 gpio controller. So it works in general.

However, when I run debian in an LXC container, with docker installed, the device is not detected anymore.  But running `dmesg | grep tty` inside the LXC container, shows the device as being detected as `dev/ttyAMA0`.

But when I try to mount this device through docker compose, the zjs-ui container doesn't start due to "no such file or directory"

Any ideas, what could be the issue? I want to run Pimox to join the RPI into my existing proxmox cluster

Following is the section in my docker compose file
```
    devices:
       - '/dev/ttyAMA0:/dev/zwave'
```

and the dmesg output from the lxc container

```
[    1.589202] fe201000.serial: ttyAMA0 at MMIO 0xfe201000 (irq = 19, base_baud = 0) is a PL011 rev2
```

**Comment by robertsLando** [comment]
Does this happens on boot or also on runtime?
I have a feel that this is a timing problem, it could be that the device is mounted with some delay and docker starts before the mount happens? 

Did you also tried to check the `/dev/serial/by-id` path?

**Reply by Wri84** [reply]
sadly, there is no `/dev/serial/by-id/` when using the GPIO pins to connect instead of USB (at least as far as I'm aware).

I've tried again to setup zjsui through docker compose, but again am not able to connect to the the device.

Following this [guide](https://coldcorner.de/2018/07/12/proxmox-usb-passthrough-fuer-lxc-container-z-wave-uzb1/ ) (which is about passing through USB devices, but I hope works similarly for serial devices), 

i ran `ls -l /dev/ttyAMA0` which results in 

```
# ls -l /dev/ttyAMA0# ls -l /dev/ttyAMA0
crw-rw-rw- 1 root dialout 204, 64 Oct 19 14:22 /dev/ttyAMA0
```

I've added the following arguments to my lxc.conf file for the zjsui debian container:

```
lxc.cgroup.devices.allow: c 204:* rwm
lxc.mount.entry: /dev/ttyAMA0 dev/ttyAMA0 none bind,optional,create=file
```

and afterwards changed the permissions on `/dev/ttyAMA0` as follows

`chmod o+rw /dev/ttyAMA0`

I changed the provided docker-compose file to bind `dev/ttyAMA0` to `/dev/zwave` and the container boots up fine, allowing to connect to the web interface. However, here I receive the following error for the device

`Driver: Failed to open the serial port: Error: Operation not permitted, cannot open /dev/zwave (ZW0100)`

The LXC container is run in priviliged mode. 

Is this a problem with the permissions?

**Reply by Wri84** [reply]
just to add:

- When adding the user to the dialout group, I can't connect at all anymore. 
- when adding GID=20 in the environment variables (with 20 being the dialout group), I still receive the same error

**Reply by Wri84** [reply]
and one more development.

I've changed the cgroup to this

lxc.cgroup**2**.devices.allow: c 204:* rwm

and now receive the following error in zjsui

`Driver: Failed to open the serial port: Error Resource temporarily unavailable Cannot lock port (ZW0100)`

All I can gether from my searches is that others had this problem, when 2 instances run at the same time. But I don't do that, its only the docker container acessing the device.

Any hints where to look?

**Reply by robertsLando** [reply]
Unfortunately I have 0 experience with Proxmox and LXC containers so I really dunno how to help here, I tried to google something and I found this:

https://tteck.github.io/Proxmox/

If you expand `Zigbee - Zwave` > `Z-Wave JS UI LXC` there is a script that should do the trick

Alternatively I also found this: https://github.com/Koenkk/zigbee2mqtt/issues/7380

Different project but same issue

**Reply by lorenzothurman** [reply]
Thank you, thank you, thank you! I've been searching high and low trying to get my zwave/zigbee controller to pass through to PiMox. Finally, thanks again

**Comment by Wri84** [comment, marked as answer]
Success! 

Thanks @robertsLando for pointing me to the same issue for zigbee, however this was for USB devices as well, so it took some time for me to finally solve it.

The Aeotec guide for the Z-Pi7 is helpful to get the gpio board available on the raspberry pi baremetal, passing it through to an LXC container needs the following steps:

1. One needs to create a udev rule in `/etc/udev/rules.d` to persist write access and passthrough the user group. Mine looks as follows to passthrough the `ttyAMA0` port

```
SUBSYSTEMS=="tty", ENV{ID_PATH}=="/devices/platform/soc/fe201000.serial/tty/ttyAMA0", SYMLINK+="ttyACMZwave", GROUP="dialout", MODE="0666"
```
2. the `/etc/pve/lxc/[VM-ID].conf` file needs the mount point added. I put in the following at the end:

```
lxc.cgroup2.devices.allow: c *:* rwm
lxc.mount.entry: /dev/ttyAMA0 dev/ttyACMZwave none bind,optional,create=file
```
(I think by allowing `*:*` the container may now access everything, but at least it works). 
Important parts are `cgroup2` for Proxmox 7.x, whereas it was just `cgroup` before

3. Reboot the Pi
4. Finally, adjust the docker-compose file to use the `/dev/ttyACMZwave` device. 

All guides were just discussing USB devices and using `/dev/serial/by-id/` paths, which are not created on the Rpi for the GPIO serial devices. 

So now I'm able to have my Z-Wave-Js-UI container running inside an LXC container on Pimox, while passing through the Aeotec Z-Pi 7 GPIO board, with Homeassistant now inside a VM on another host and connected over LAN.

Thanks again for the help and hopefully this will help someone in the future :)

**Reply by robertsLando** [reply]
Thanks for the feedback and guide @Wri84 🙏🏼
